Save Navigation Block data to a wp_navigation post type

talldan opened the pull request:
## Description
Closes #34612. Alternative to https://github.com/WordPress/gutenberg/pull/35418, that instead of using a template part post type, uses a dedicated `wp_navigation` custom post type for saving the navigation block's inner blocks.

TODO:
- [x] Add a switcher to the block toolbar for choosing another navigation post.
- [x] Renaming of navigation posts.
- [x] Decide what to do about the wp-admin view of navigation posts. Maybe hide it and then follow up with work on a way to edit in isolation? (edit: have disabled this, but it leaves an issue in that there's no way to delete menus)
- [x] Improve copy. What should we publicly call the user facing post type for a navigation block? Is 'Navigation blocks' ok, or is that confusing? (edit: I've settled on 'Navigation Menu' for now)
- [x] Handle existing navigation blocks in content. Offer a path for saving their inner blocks?
- [x] Fix any failing tests

## How has this been tested?
#### New blocks
1. Open the site editor
2. Add a navigation block
3. Choose 'new menu'
4. Add some blocks
5. Save (making sure to save the navigation data) and reload, content should still be present
6. Add a new navigation block elsewhere
7. Select the previously created menu
8. Modify the inner blocks, both navigation blocks should show the updated content
9. Apply styling to one of the blocks. Only the modified block should be affected.

#### Upgrading blocks
1. Checkout `trunk`, build the codebase.
2. In the site or post editor, create a nav block with some inner blocks and save the content.
3 Checkout this branch
4. Follow the process for upgrading the block

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

talldan committed:
Explore using a template part for the navigaiton block data

Destructure attributes in nav block

Add use of template part entity

Move edit into a folder like the template part block

Initial template part selection

Render the template part inner blocks

Fix visibility of UI when placeholders are visible

Remove use of theme attribute

Rework placeholder

Revert "Remove use of theme attribute"

This reverts commit 2adbdf7a7d833acddd498d32b7069322e4514522.

talldan committed:
Register custom post type

talldan committed:
Save data to a navigation post type

Fix showing existing navigation post inner blocks

jasmussen commented:
Thanks for taking a stab at this. I'm tripping myself on some of the testing steps, and find myself unable to select a previously created menu. What am I doing wrong in this GIF?
![nav](https://user-images.githubusercontent.com/1204802/137873569-096ddcc5-f926-4912-9382-d9b21af9579d.gif)

talldan commented:
@jasmussen That's interesting. You should be seeing a two-step block placeholder, first step should look like this, similar to when a template part is added:
<img width="753" alt="Screenshot 2021-10-19 at 4 39 34 pm" src="https://user-images.githubusercontent.com/677833/137874695-215e9c01-6ef7-40e4-8892-c787b1bfc4f4.png">

talldan commented:
@jasmussen That might happen if your code isn't re-built for some reason, maybe the build failed? It looks like you have the PHP changes but not the JavaScript changes.

jasmussen commented:
Thanks, I'll give this another shot. Back in a moment.

jasmussen commented:
This is an impressive PR:

![test](https://user-images.githubusercontent.com/1204802/137876966-5af9a97d-6ed2-4afb-981d-d5c96b6144d7.gif)

Whether data should be saved as a custom post type or as a template part, I'm unqualified to chime in on. Focusing purely on the navigation block part of it, the primary effort would be to integrate a menu picker into the existing setup state so we end up with a one-step placeholder. I'll note down a task to do some mockups here.

draganescu commented:
Took this for a spin and just like @jasmussen I am amazed how impressive it is. I did not have to figure it out at all, it just worked. The one problem I found is that the front end does not render the menu I see in the Site Editor. Is this known or am I doing it wrong?

I see this in the editor:

<img width="1206" alt="Screen Shot 2021-10-19 at 17 07 14" src="https://user-images.githubusercontent.com/107534/137926874-a0d30b5d-eadc-4cf0-bdf1-1a77cfdc9422.png">

And this rendered:

<img width="1223" alt="Screen Shot 2021-10-19 at 17 07 23" src="https://user-images.githubusercontent.com/107534/137926906-00c47282-c2f9-4094-9660-edc16f80c6b2.png">

gwwar commented:
This one looks pretty interesting @talldan! Nice work. ✨ 

#### Things that need fixing

❌ From testing I believe top level attributes on the nav block like vertical variation aren't being saved. Would it be simpler to insert a block reference, and have the navigation CPT store both the nav block and the inner-content? 

#### Context on Flow

So for folks reviewing, this PR plays around with the serialization of the block. Instead of storing link content inline, we instead create a nav CPT, and simply link to the post id:

| Nav block points to the CPT id | CPT content stores the inner content (everything but the nav container) | If we edit 3 Nav blocks, we see these queued up for save on publish |
|------|------|------|
| <img width="1119" alt="CleanShot 2021-10-19 at 10 37 44@2x" src="https://user-images.githubusercontent.com/1270189/137963400-2ea235bd-7f75-41c9-835c-eb04a5decc0d.png"> | <img width="1116" alt="CleanShot 2021-10-19 at 10 37 24@2x" src="https://user-images.githubusercontent.com/1270189/137963342-a9cd1aed-6a8a-4ae0-a929-cee999d65a2b.png"> | <img width="297" alt="CleanShot 2021-10-19 at 10 48 00@2x" src="https://user-images.githubusercontent.com/1270189/137963799-9325d522-baef-4223-b98b-03303cd5a23f.png"> |
 
In terms of workflow, there's a new step when adding a nav block, we give it a name.  I understand that we're still prototyping here. Some suggested feedback for the new flow:

* To make this a little friendlier we might want to suggest something more memorable to start as a placeholder prompt like "Header menu" or "Nav menu" etc. Since each is backed by a CPT, it'd be ideal to also show the navigation menu name in ListView and when we select the block. (Similar to how we show named headers).
* Step 2 likely needs some copy changes or some 🤔 After naming a menu it's confusing to pick from other named menus. Maybe something like populate, or start from? Can we also start from content from other CPT nav menus?
* More of an enhancement for another PR but I wonder if at step 2 we can show some pattern placeholders w/hover preview. It'd make this step less intimidating.

![CleanShot 2021-10-19 at 10 38 09@2x](https://user-images.githubusercontent.com/1270189/137964742-b3b33b56-9a9e-477b-9b6d-474b4144c9f4.png)

https://user-images.githubusercontent.com/1270189/137963106-f1ace6a0-e3ba-4638-bd63-afc63abbe1f2.mp4

#### Questions from summary

> Handle existing navigation blocks in content. Offer a path for saving their inner blocks?

There's a lot of similarities here for saving navigation content with a CPT and reusable blocks. Provided we are pretty clear on what is backed by a CPT and what's saved inline to a post, I think it might be a simple as showing some actions like "convert to reusable menu" and "convert to regular blocks" (or whatever we decide to call these sorts of nav blocks).

<img width="405" alt="CleanShot 2021-10-19 at 11 00 09@2x" src="https://user-images.githubusercontent.com/1270189/137965547-5ef6613f-f4d8-4429-8d39-bf6ed2cb7b08.png">

> Decide what to do about the wp-admin view of navigation posts. Maybe hide it and then follow up with work on a way to edit in isolation?

Probably a good opportunity to edit these CPTs with a block based editor like the nav editor. The post editor is serviceable enough in the meantime, but folks might need some explanation of what it is. We could maybe add a filter for this CPT, and add a wrapper nav block that doesn't get serialized? Or perhaps save the entire nav + contents in one place.

> What should we publicly call the user facing post type for a navigation block?

There's already "Menus" so how about "Block Menus"? No strong feelings on this one.

#### Misc thoughts

One thing that comes to mind is that if we use a CPT, there's very little difference here between backing a nav block with a CPT vs one with an existing menu. The latter would likely need a simple adapter, and some extra logic limiting the type of inner blocks we can add.

Overall, I think the idea here is pretty promising. It be great to get some feedback from folks more familiar with how folks prefer building sites. Perhaps a +1 review from folks more familiar with themes, and another from APIs?

tellthemachines commented on lib/navigation.php:
```diff
@@ -384,3 +384,57 @@ function gutenberg_add_block_menu_item_styles_to_nav_menus( $hook ) {
+function gutenberg_register_navigation_post_type() {
+	// TODO - Some of the language here needs to be revised. 'Navigation' is a
+	// singular noun, so cannot comfortably be used in a plural form, which
+	// leads to a situation where something like 'menus' needs to be suffixed.
```
Could we call the post type "menu" instead?

getdave replied:
In general can we decide on "navigation" vs "menu". I had understood we were normalizing towards "navigation" so as to disambiguate from classic Menus.

talldan replied:
@getdave Navigation is a singular noun, it doesn't have a plural form, so I think we're going to struggle to use it everywhere. Have a look through the strings below this comment to see what I mean.

talldan replied:
I went with Navigation Menu for now, it can be refined.

tellthemachines commented on lib/navigation.php:
```diff
@@ -384,3 +384,57 @@ function gutenberg_add_block_menu_item_styles_to_nav_menus( $hook ) {
+		'insert_into_item'      => __( 'Insert into Navigation', 'gutenberg' ),
+		'uploaded_to_this_item' => __( 'Uploaded to this Navigation', 'gutenberg' ),
+		// Some of these are a bit weird, what are they for?
+		'filter_items_list'     => __( 'Filter Navigation list', 'gutenberg' ),
```
This one seems to provide the screen reader text for the "Filter" section in the Navigation post type index page:

<img width="1682" alt="Screen Shot 2021-10-20 at 5 23 03 pm" src="https://user-images.githubusercontent.com/8096000/138039193-969bc9a4-8226-4172-8f9a-8bb6d7dd266a.png">

The docs for `register_post_type` might need updating 😅

talldan replied:
Thanks for finding that!

tellthemachines commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -231,37 +188,49 @@ function Navigation( {
-					} }
-				/>
+				<Warning>
+					{ __( 'Block cannot be rendered inside itself.' ) }
```
I was still able to add the block inside itself if I go through "All Navigation Menus" to the standalone editor. It rendered the whole block, with the warning in the place where the nested block would be. Then it started flickering, but I was able to save it.

Interestingly enough, when I tried the same with a menu that wasn't being used in any location, it didn't show that menu as an option to pick from in the site editor. But if the menu is already being used somewhere, it still shows.

getdave replied:
I assume that's because the isolated editor doesn't have a _wrapper_ Nav block. There's work to do there in general about allowing insertion of _child_ Nav blocks outside of the Nav parent block.

talldan replied:
Yeah, the standalone editor is just a post editor.

> Interestingly enough, when I tried the same with a menu that wasn't being used in any location, it didn't show that menu as an option to pick from in the site editor. But if the menu is already being used somewhere, it still shows

@tellthemachines would you be able to elaborate on this?

tellthemachines replied:
What I did was:

- Create a new nav post from the "All Navigation Menus" page.
- Give it a name and populate it only with blocks that can exist inside a Nav, such as Search and Social links.
- Save the post. I think I refreshed the page after that.
- Add a Navigation block to the post, click "Choose existing" and choose the one I am currently editing.
- The Navigation was added to the post, and inside it were all the blocks I had added above, plus the warning that the block can't render inside itself.
- I then went to to the Site Editor, and added a new Nav block to the index template.
- Under "Choose existing", the navigation that I had just created did not appear.
- I then went back and edited the post to remove the nested Nav block.
- Going back to the Site Editor and refreshing the page, that post now appeared under "Choose existing" in my new nav block.

Then I tried:

- From the "All Navigation Menus" page, opening an existing nav post that was already in use in a template.
- Adding a new nav block inside it, and choosing itself from "Choose existing".
- The block was added, but immediately started flickering, and clicking "update" made the whole editor hang, though it did save the post after a while. (The flickering and hanging didn't happen in the previous test.)
- I went to the Site Editor, and added a new Nav block to the index template.
- Under "Choose existing", I was able to choose the navigation that I had just mangled by rendering it inside itself.
- Both that, and the other existing instance of the same navigation in that template, displayed the warning about the block rendering inside itself, in addition to the other blocks that nav contained.

talldan replied:
Yeah, the editor at 'All Navigation Menus' is just the post editor, and as @getdave mentioned, because there's no wrapping Navigation Block, there's no recursion provider in that view, so it's possible for a user to create anything there.

I think for this PR, that option should be hidden in WP Admin.

The flickering sounds unusual. Not sure why that would happen.

You also mentioned this:
> Under "Choose existing", the navigation that I had just created did not appear.

At the moment, drafts don't appear in that menu, so it might be that you saved it as a draft? I couldn't reproduce what you mentioned otherwise.

tellthemachines reviewed (commented):
This is looking really good! 

In addition to the todoes other folks have mentioned (block not rendering on the front end and not saving customizations such as color and variation jumped out the most to me), there's also:

* direct insert logic not working as expected (it always assumes direct insert is true, independently of what type of blocks are in the nav);
* in editor accessible from "All Navigation Menus", all blocks can still being inserted, as if it were a regular post. (I like the idea of leveraging the navigation editor for this view, but it doesn't feel like a must-have for V1 of this work, so perhaps we can just hide it for now.)

As for the CPT vs Template Part, there doesn't seem to be too much duplication of logic in this PR? I reckon it's worth going with the CPT.

➕ to making the 2-step placeholder flow 1 step only 😄

getdave reviewed (commented):
Great work exploring these PRs Dan. Very much appreciated 🙇 👏 👏 👏 

I've undertaken a [review of this on video](https://d.pr/v/66JcKN) (with audio!).

## What are we looking to achieve?

Before I make comments let's reflect on the wider goals we're looking to address (some of which may not be addressed in this PR):

* Allow navigations to be used in different locations within the same template/site.
* Allow reuse of the same navigation _data_ but with different _presentational_ treatments.
* Retain ability to quickly build new navigations.
* Separate presentation and data in order to afford editing navigation data in isolation (e.g. Nav Editor project).
* Allow reuse of navigations across Themes.


Overall, I think this PR is addressing (with some refinement) the first 4 items which is great (the point about Theme switching is being discussed in https://github.com/WordPress/gutenberg/issues/35750).


## Notes from Video 

- The workflow was a lot more intuitive - less baggage from Template Parts is a win.
- I tried using the same menu in the header and footer but applying _different_ presentational treatments (text colors...etc). This worked initially but when I reloaded some of these visual distinctions were lost.
- I really like the isolated view which has just the nav items and not the wrapper. This for me is where the Nav Editor should be headed. Avoids all baggage from the Nav block and provides a very useable experience to manipulate the _data_.
- as in https://github.com/WordPress/gutenberg/pull/35418 when viewing the CPT in isolated view you cannot add Nav block items because they are not part of parent. We should be able to work around that somehow.


## Technical approach

I would also like to emphasise that I believe you are taking the _correct_ route in _not_ serializing the Nav _block_ itself to the CPT. The block should be a wrapper concerned solely with presentation specific to the location (within the template) in which it is being used, where as the inner blocks should represent the data structure of the navigation itself. If we were to serialize the Nav block then we'd also persist all the presentation and it would therefore make it far less flexible when attempting to reuse the same navigation _data_ between instances (eg: header vs footer) or indeed within an Isolated Editor view (e.g. Nav Editor).

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -6,70 +6,40 @@ import classnames from 'classnames';
-	type: 'default',
-	alignments: [],
-};
+// TODO - refactor these to somewhere common?
```
Such as a hypothetical `@wordpress/navigation` or `@wordpress/menus` or within the block "codebase" itself?

talldan replied:
I think this comment is also left over from the template part implementation, it can be deleted.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -231,37 +188,49 @@ function Navigation( {
-			? CustomPlaceholder
-			: NavigationPlaceholder;
+	// We don't want to render a missing state if we have any inner blocks.
+	// A new template part is automatically created if we have any inner blocks but no entity.
```
Do we need to update this comment as it's referring to "template part"?

talldan replied:
Yep

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -231,37 +188,49 @@ function Navigation( {
-			? CustomPlaceholder
-			: NavigationPlaceholder;
+	// We don't want to render a missing state if we have any inner blocks.
+	// A new template part is automatically created if we have any inner blocks but no entity.
```
```suggestion
	// If we have a reference to the Navigation Post but the post entity no longer exists
	// then show a warning to the user.
```

Question: if the users ends up in this state can we provide them with an option to create a new navigation rather than having to remove the block and re-add it?

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -231,37 +188,49 @@ function Navigation( {
+						__(
+							'Navigation block has been deleted or is unavailable: %s'
+						),
+						navigationPostId
```
Not sure how useful an ID is in a user-facing error but I appreciate we aren't storing any other attributes of the navigation CPT to the block so you can't use the `name` or `title` or whatever.

talldan replied:
Yeah, agreed. Maybe we should consider using a slug for Navigation Posts.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -372,65 +341,49 @@ function Navigation( {
+						onFinish={ ( post ) => {
+							setIsPlaceholderShown( false );
+							setAttributes( {
+								navigationPostId: post.id.toString(),
```
I didn't appreciate it would be necessary to store as a string here. Do we not want a true number?

talldan replied:
Yeah, I think so. It was something that I fixed as quickly as possible and not in the best way in order to get the PR up for feedback. I'll improve this.

getdave commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -0,0 +1,83 @@
+	const { saveEntityRecord } = useDispatch( coreStore );
+
+	const createNavigationPost = useCallback(
+		async ( title = __( 'Untitled Menu' ), blocks = [] ) => {
```
```suggestion
		async ( title = __( 'Untitled Navigation' ), blocks = [] ) => {
```

getdave commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -0,0 +1,83 @@
+			// If we have `area` set from block attributes, means an exposed
+			// block variation was inserted. So add this prop to the template
+			// part entity on creation. Afterwards remove `area` value from
+			// block attributes.
```
Is this comment carried over from the Template Parts exploration?

talldan replied:
Yeah, I need to tidy some bits up. 👍

getdave commented on packages/block-library/src/navigation/edit/placeholder/select-navigation-post-step.js:
```diff
@@ -0,0 +1,133 @@
+				instructions={
+					canSwitchNavigationPost
+						? __( 'Choose an existing menu or create a new one.' )
+						: __( 'Create a new menu.' )
```
```suggestion
						? __( 'Choose an existing navigation or create a new one.' )
						: __( 'Create a new navigation.' )
```

getdave commented on packages/block-library/src/navigation/edit/placeholder/select-navigation-post-step.js:
```diff
@@ -0,0 +1,133 @@
+					</>
+				) }
+			</Placeholder>
+			{ isNewMenuModalVisible && (
```
Instead of a modal here, why don't we allow simply entering a menu title inline within the placeholder?

talldan replied:
It's a pattern in the reusable block and template part. I think it's fine to keep it like that right now.

I think as a follow-up to this the whole placeholder flow will have to be revised.

spacedmonkey commented:
Question: If we store menus in this new post type, will titles of posts, pages etc will be dynamitically be updated like the current menus. Will pages / posts that are deleted, automatically be removed from this content? Will links to CPT that are no longer be register also be removed. 

Menus have be somewhat dynamic, as they link to other WordPress content. If menu is not dynamic, then this is a massive step backwards for users, IMO.

getdave commented:
https://user-images.githubusercontent.com/444434/138108047-d15609c5-2525-4c5a-bce1-7865b70f0afd.mov

Another cool knock on effect of this direction is that you can now make changes to your navigation without causing changes in the associated template part.

So say for example, you have a navigation in your `header` template part, you can now add a new item to the navigation and it won't cause the `header` to be marked as "dirty" (ie: it won't need saving). Only the navigation itself will need changing.

cc @fabiankaegy

getdave commented:
> Question: If we store menus in this new post type, will titles of posts, pages etc will be dynamitically be updated like the current menus. Will pages / posts that are deleted, automatically be removed from this content? Will links to CPT that are no longer be register also be removed.
> 
> Menus have be somewhat dynamic, as they link to other WordPress content. If menu is not dynamic, then this is a massive step backwards for users, IMO.

I [gave this a test](https://d.pr/v/PrbiIV/YlSNu62ATJ) and it _doesn't_ seem to remove the page from the navigation if I remove the page from `Pages`.

~I wonder if the link item blocks are storing a _reference_ to the page they were created from or simply inlining the details of the page (eg: url, title...etc) that they receive _at the point of creation_~ 🤔  I checked and they seem to hardcode everything rather than store a reference

<img width="816" alt="Screenshot 2021-10-20 at 15 21 54" src="https://user-images.githubusercontent.com/444434/138111708-a781c45a-a951-441d-bc61-ba3e58fdfe57.png">


This seems more a limitation of the Nav Link block rather than the CPT approach in general. Perhaps the Nav link block should only store a reference to the entity it represents. Then when we grab the blocks from the Nav CPT we could request details for each using the reference (probably a post id) prior to rendering. If the entity represented by the block no longer exists then we show an error state. This will also ensure URLs are automatically updated to reflect the underlying entity.

This only applies to Nav items which represent entitities. Custom links (eg: `www.wordpress.org`) don't need to worry about this.

talldan commented:
Thanks for all the feedback!

I think there are two main bugs that are important to solve right now:
- The issue mentioned where navigation block attributes aren't being saved (colors, font styles).
- The block not rendering correctly on the front-end. (270ef0e fixes this)

The second one seems fairly obvious to me. I just hadn't looked at this part of the block yet 😄 

I'll look at the first bug as well, but if anyone is able to lend a hand with that I'd really appreciate the assistance.

The other two bits of common feedback:
- The placeholder flow could now use refinement - lets push this to a separate issue/pr and get some design iteration happening. (https://github.com/WordPress/gutenberg/issues/35818)
- The  post editor for each menu available from 'All Navigation Menus' doesn't really work correctly - again, I think it'd be best to explore this in a follow-up, I'll disable it for now here.

talldan commented:
> Question: If we store menus in this new post type, will titles of posts, pages etc will be dynamitically be updated like the current menus. Will pages / posts that are deleted, automatically be removed from this content? Will links to CPT that are no longer be register also be removed.

> Menus have be somewhat dynamic, as they link to other WordPress content. If menu is not dynamic, then this is a massive step backwards for users, IMO.

@spacedmonkey I don't think this PR would change anything related to that. The Navigation Link block is a dynamic block, so it could already support most of these features, they just haven't been implemented or considered yet.

<strike>The only part that behaves like you describe currently is the URL for a link, I believe that should update to reflect permalink changes.</strike> (edit: it doesn't, see https://github.com/WordPress/gutenberg/issues/18345).

I think it'd be good to create issues for those other things 👍 

edit: actually I double-checked this and a link is no longer rendered on the frontend of a site if an associated post/page is removed. In the editor, there's still a link block present, I think ideally the user would be prompted to update/remove this block.

talldan committed:
Output blocks from Navigation Post on frontend

talldan committed:
Avoid rendering block when no navigation post is set

talldan commented:
> I tried using the same menu in the header and footer but applying different presentational treatments (text colors...etc). This worked initially but when I reloaded some of these visual distinctions were lost.

So I've discovered this is also a bug on the template part block. To reproduce try this:
1. Add a template part block to an empty post in the post editor
2. Choose an existing template part from the placeholder
3. Apply padding, change colors
4. Switch to code editor mode. Observe none of the attributes for colors or padding are serialized in the block delimiter.
5. Switch back to visual editor mode.
6. Add a block after the template part (e.g. a paragraph with a single character)
7. Switch to code editor mode. See that there are now color and padding attributes.

Also, the following steps are curious:
1. Add a template part block in the post editor
2. Apply padding, change colors
3. Switch to code editor mode. Observe the attributes for colors or padding are serialized in the block delimiter.

It seems like the existence of inner blocks causes an issue. I have a feeling this might be a problem with `useEntityBlockEditor` and perhaps nested entities. It seems like changes to the block wrappers are not causing the post content to be edited. The changes are shown in the block editor as the block editor store has the correct state.

Possibly related to https://github.com/WordPress/gutenberg/issues/35664? From the sounds of it, the bug might have been introduced in 11.5 - https://github.com/WordPress/gutenberg/releases/tag/v11.5.0

I did a git bisect and managed to track the issue down to #34241.

draganescu commented:
cc @spacedmonkey 

> Question: If we store menus in this new post type, will titles of posts, pages etc will be dynamitically be updated like the current menus. Will pages / posts that are deleted, automatically be removed from this content? Will links to CPT that are no longer be register also be removed.

By default no. In this PR we're only starting by making block navigation data reusable in other navigation blocks. To achieve:

- automatic page addition and removal from menus
- automatic menu title updates for page title changes
- automattic CPT addition and removal from menus

... new PRs building on this one will be needed.

> If menu is not dynamic, then this is a massive step backwards for users, IMO.

As @talldan mentioned, the navigation block's inner workings will be dynamic where this is needed. What will come about in near future iterations is creating the mechanisms to achieve a comparable level of functionality. For example, in the block navigation we have the pages list block that is self updating with new pages / page titles. This is not a feature of the navigation block, but a feature of the page list block. So the exact way of implementation will have to be adapted to the block foundation that we're working with.

talldan committed:
Fix direct insert

talldan committed:
Update List View feature

talldan committed:
Add menu switcher

talldan committed:
Fix types

talldan committed:
Use PlaceholderPreview when loading

talldan committed:
Add menu naming

talldan committed:
Update PHP strings

talldan committed:
Hide post editor view

talldan committed:
Update more PHP strings

talldan committed:
Rename navigation post to navigation menu

talldan commented:
## Progress update
This should now be working much better, and I think is close to being ready. I've fixed all of the bugs mentioned apart from the one about saving style attributes, which is a bug in `trunk` and being explored here - https://github.com/WordPress/gutenberg/pull/35827.

- Fixed front-end rendering of block
- Fixed direct insert
- Settled on the name 'Navigation Menu' for the post.
- Hid the wp-admin access to the post editor for Navigation Menus until that's resolved
- Improvements to the loading state of the placeholder
- Added a toolbar to switch to a different Navigation Menu when one is already selected
- Added a way to rename Navigation Menus in the block inspector
- Added a way to upgrade old Navigation Block content that doesn't use the Navigation Menu CPT

## TODO
- Fix tests
- Anything else that anyone can spot?

## Follow-ups
- Refine placeholder flow (https://github.com/WordPress/gutenberg/issues/35818)
- Explore isolated Navigation Menu editing (needs an issue)
- Explore storing a JSON representation of menus in the CPT (https://github.com/WordPress/gutenberg/issues/35868)
- Fix controlled block attribute saving (https://github.com/WordPress/gutenberg/pull/35827)
- Explore theme switching (https://github.com/WordPress/gutenberg/issues/35750)

talldan committed:
Add upgrade process

talldan commented:
I've added a way to upgrade existing navigation block content to use the Navigation Menu entity:
<img width="359" alt="Screenshot 2021-10-22 at 3 43 15 pm" src="https://user-images.githubusercontent.com/677833/138414831-fd5721b2-7f6d-40c8-af3d-8e7154799e6c.png">

Copy could probably do with some refinement.

talldan committed:
Show menus in ui, but remove editor support and fix some labels

talldan committed:
Fix empty block edit if block uses an empty entity

talldan committed:
Hide wp-admin CPT view again

talldan committed:
Fix frontend rendering from menu entity

talldan committed:
Preserve inner blocks if post content is saved while upgrade warning is shown

talldan committed:
Skip e2e tests

talldan commented:
I've decided to skip the Navigation Block e2e tests for now.

They generally need some attention. So I think I'll try a rewrite next week of the tests using some of the learnings from the nav editor, which had very stable tests and moved on from using mocks.

The other issue is that the block now serializes to something like `<!-- wp:navigation {"navigationMenuId":434} /-->`, we can no longer assert that the inner blocks are correct in the same way. So that will require some revamping.

I'll also have to disable the nav editor tests as that no longer works.

talldan committed:
Disable navigation editor tests

gwwar commented on packages/block-library/src/navigation/index.php:
```diff
@@ -238,16 +239,35 @@ function render_block_core_navigation( $attributes, $content, $block ) {
+
+		// For some reason 'parse_blocks' includes a null block with '\n\n' as the content?
+		// This code strips it.
+		$compacted_blocks = array_filter(
```
🤔 It feels like we shouldn't need to remove items. I'm also seeing an empty block item between links. 

Removing the whitespace fixes it. Are we missing any filters that need to be run on the post content, before calling parse_blocks?

Perhaps @dmsnell would know? For context, this PR experiments with storing navigation links inside of a custom post type. The wrapping nav block is not persisted in the CPT for general link reuse. Eg a nav block on the front page may set different nav block params, but use the same backing links. 

Here's an example debug print:

![CleanShot 2021-10-22 at 15 15 10@2x](https://user-images.githubusercontent.com/1270189/138528695-5325e073-e9b9-462f-adbd-65dca513ed10.png)

dmsnell replied:
The first-stage of parsing converts HTML documents into Gutenberg documents in an isomorphic way, which means that excess whitespace is preserved as raw HTML. That's what those "empty" blocks are. There are parser filters that can be used to eliminate these so that they don't have to happen through `array_filter()`, but `array_filter()` seems fine. The cause for this awkward convention is beauty and peoples' preference to add newlines between blocks in the serialized HTML.

Cf. #34023

gwwar replied:
> which means that excess whitespace is preserved as raw HTML

Oh interesting. TIL! Thanks for clarifying @dmsnell

gwwar reviewed (commented):
I'm still seeing no navigation attributes being serialized, so things like colors, text styles and variations don't work on the published view. Did I miss any additional testing instructions?

tellthemachines commented:
> edit: actually I double-checked this and a link is no longer rendered on the frontend of a site if an associated post/page is removed. In the editor, there's still a link block present, I think ideally the user would be prompted to update/remove this block.

#23573 and its PR #31716 should address this.

tellthemachines reviewed (commented):
It's great to see this working so well 🎉 Feels almost there!

A couple of things I noticed in this round of testing:

* Nav blocks seem to take a bit longer to render when loading in both post and site editors; on trunk they load straightaway.

* In the post editor, editing a nav block and saving now shows the "Select the changes you want to save" dialog, with changes to the nav block appearing separate from changes to the post. I'd expect this when changing the contents of the block, but not when making a presentational change e.g. background color, because that should apply only to this instance of the block, not to the CPT content. Right now it's appearing even if changes are only presentational.

talldan commented:
Thanks for the feedback!
> Nav blocks seem to take a bit longer to render when loading in both post and site editors; on trunk they load straightaway.

Hmm, yeah, there's some asynchronicity from the REST request needed to load the navigation block data. This will be a difficult one to solve, but I'll look at how it can be improved. Some form of preloading would be good, but the server side code would need to know what to preload.

> In the post editor, editing a nav block and saving now shows the "Select the changes you want to save" dialog, with changes to the nav block appearing separate from changes to the post. I'd expect this when changing the contents of the block, but not when making a presentational change e.g. background color, because that should apply only to this instance of the block, not to the CPT content. Right now it's appearing even if changes are only presentational.

This one is caused by a separate editor bug that will be important to fix for 5.9. It's currently an issue with template parts too. This is the issue for it - https://github.com/WordPress/gutenberg/issues/35664, and there's a fix in the works in https://github.com/WordPress/gutenberg/pull/35827, but it's a complex one to get right.

talldan committed:
Tidy up comment

talldan commented:
> Nav blocks seem to take a bit longer to render when loading in both post and site editors; on trunk they load straightaway.

> Hmm, yeah, there's some asynchronicity from the REST request needed to load the navigation block data. This will be a difficult one to solve, but I'll look at how it can be improved. Some form of preloading would be good, but the server side code would need to know what to preload.

@tellthemachines This is going to be a difficult one to improve. I think it needs more discussion. I've made an issue outlining the challenges and detailing a potential solution - https://github.com/WordPress/gutenberg/issues/35942

noisysocks commented:
Nice work! Let's get this in early and iterate since other issues and PRs depend on it. I added "needs dev note" since it would be good to explain the CPT intricacies on Make/Core.
